src/Hazelcast.Net/CP/ClusterCPGroups.cs

```diff
@@ -0,0 +1,131 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+// http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Hazelcast.Clustering;
+using Hazelcast.Core;
+using Hazelcast.Exceptions;
+using Hazelcast.Models;
+using Hazelcast.Protocol.Codecs;
+using Microsoft.Extensions.Logging;
+namespace Hazelcast.CP
+{
+    /// <summary>
+    /// Manages CP group information in the cluster.
+    /// </summary>
+    internal class ClusterCPGroups
+    {
+        private long _version = InitialVersion;
+        private readonly object _mutex = new();
+        private readonly object _listenerMutex = new();
+        private readonly ConcurrentDictionary<CPGroupId, Guid> _groups = new();
+        private readonly ILogger _logger;
+        private readonly ClusterState _clusterState;
+        private readonly ClusterMembers _clusterMembers;
+        private int _disposed;
+        private MemberConnection _listenerConnection;
+        private Task _listenerRegisterTask;
+        private long _correlationId;
+
+        private CancellationTokenSource _cts;
+
+        public const string CPGroupsJsonField = "cp.leaders";
+        public const long InitialVersion = -1;
+
+        public ClusterCPGroups(ClusterState clusterState, ClusterMembers members)
+        {
+            _clusterState = clusterState;
+            _clusterMembers = members;
+            _logger = _clusterState.LoggerFactory.CreateLogger<ClusterCPGroups>();
+        }
+
+        public long Version => _version;
+        public int Count => _groups.Count;
+
+        public void SetGroups(long version, ICollection<CPGroupInfo> groups, IList<KeyValuePair<Guid, Guid>> cpToApUuids)
+        {
+            if (_version >= version)
+                return;
+
+            var oldCount = _groups.Count;
+
+            var newGroup = MapCPtoAPUuids(cpToApUuids, groups);
+
+            lock (_mutex)
+            {
+                // One more check in safe env.
+                if (_version >= version)
+                    return;
+
+                _version = version;
+
+                _groups.Clear();
+                foreach (var g in newGroup)
+                {
+                    _groups[g.Key] = g.Value;
+                }
+
+            }
+
+            _logger.IfDebug()?.LogDebug("CP groups updated to Version {Version}." +
+                                        " Old groups count: {OldCount}, new count: {NewCount}", Version, oldCount, groups.Count);
```

Comment:
Is it a frequent log message? If it's really infrequent then it might be worth bumping to `INFO` as nobody is running with `DEBUG`. If it's frequent then I would leave as-is.